src/particle_filter/src/particleFilter6DOF.cpp

```diff
@@ -0,0 +1,359 @@
+#include <string.h>
+#include <iostream>
+#include <random>
+#include <cmath>
+#include <chrono>
+#include "distanceTransform.h"
+#include "particleFilter6DOF.h"
+#include "matrix.h"
+using namespace std;
+
+# define Pi          3.141592653589793238462643383279502884L
+
+#define SQ(x) ((x)*(x))
+#define bzero(b,len) (memset((b), '\0', (len)), (void) 0)
+
+particleFilter::particleFilter(int n_particles, cspace b_init[2],
+			       double Xstd_ob, double Xstd_tran,
+			       double Xstd_scatter, double R)
+  : N(n_particles), Xstd_ob(Xstd_ob), Xstd_tran(Xstd_tran),
+    Xstd_scatter(Xstd_scatter), R(R), firstObs(true)
+{
+  memcpy(b_Xprior, b_init, 2 * sizeof(cspace));
+  memcpy(b_Xpre, b_Xprior, 2 * sizeof(cspace));
+
+  X = new cspace[N];
+  bzero(X, N*sizeof(cspace));
+  X0 = new cspace[N];
+  bzero(X0, N*sizeof(cspace));
+
+  createParticles(X0, b_Xprior, N);
+
+  X_1 = new cspace[N];
+  W = new double[N];
+}
+
+void particleFilter::addObservation(double obs[3], double cube[3], distanceTransform *dist_transform, int idx_obs)
```

Comment:
As earlier,
use distanceTransform &dist_transform to pass by reference, without having to deal with pointers